python/mxnet/amp/amp.py

```diff
@@ -0,0 +1,262 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+# coding: utf-8
+"""Functions for enabling AMP (automatic mixed precision)."""
+__all__ = ['init', 'init_trainer', 'scale_loss']
+
+from types import MethodType
+import logging
+import contextlib
+import numpy as np
+
+from .. import symbol
+from ..symbol import Symbol
+from ..symbol import contrib as symbol_contrib
+from .. import ndarray
+from ..ndarray import NDArray
+from ..ndarray import contrib as ndarray_contrib
+from . import lists
+from ..gluon import trainer
+from .. import optimizer as opt
+from .loss_scaler import LossScaler
+
+def _cast_symbol_NDArray(s, dtype):
+    if isinstance(s, Symbol):
+        return symbol.amp_cast(s, dtype=dtype)
+    elif isinstance(s, NDArray):
+        if (s.dtype != dtype and
+                (s.dtype == np.float16 or s.dtype == np.float32) and
```

Comment:
How about make a list of all float types supported by AMP, and then we can check with the list:
```
float_types = (np.float16, np.float32) # we can extend this list if we have other float types support
if s.dtype in float_types:
```

Reply:
Ok.